File: tools/godofprompt-scraper/extract_godofprompt.js
// ============================================================================
// GOD OF PROMPT - COMPLETE PROMPT EXTRACTION SCRIPT
// ============================================================================
// This script will extract ALL free prompts from godofprompt.ai
// Run this in your Chrome browser console (F12 → Console → Paste → Enter)
// ============================================================================

(async function GodOfPromptExtractor() {
  console.log('🚀 Starting God of Prompt Complete Extraction...');
  console.log('⏱️ Please wait - this may take several minutes...\n');

  const CONFIG = {
    baseUrl: 'https://www.godofprompt.ai/prompts?premium=false',
    maxPages: 82,
    delayBetweenPages: 2500, // 2.5 seconds between page loads
    delayBetweenClicks: 500,
    maxEmptyPages: 3 // Stop after 3 consecutive pages with no new prompts
  };

  const state = {
    allPrompts: [],
    seenUrls: new Set(),
    currentPage: 1,
    emptyPageCount: 0,
    totalPages: 0
  };

  // --------------------------------------------------------------------------
  // Helper Functions
  // --------------------------------------------------------------------------

  function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  function extractPromptsFromCurrentPage() {
    const prompts = [];
    const promptLinks = document.querySelectorAll('a[href*="/prompt?prompt="]');

    promptLinks.forEach(link => {
      const url = link.href;
      if (!url.includes('prompt?prompt=')) return;
      if (state.seenUrls.has(url)) return;

      // Find the title
      let title = 'Untitled';
      const h3 = link.querySelector('h3') || link.closest('div')?.querySelector('h3');
      if (h3) title = h3.textContent.trim();

      // Find description
      let description = '';
      const descElem = link.nextElementSibling || link.parentElement?.querySelector('div:not([class])');
      if (descElem && descElem.textContent.length < 500) {
        description = descElem.textContent.trim();
      }

      state.seenUrls.add(url);
      prompts.push({
        title,
        url,
        description: description.substring(0, 200),
        page: state.currentPage,
        extractedAt: new Date().toISOString()
      });
    });

    return prompts;
  }

  async function clickNextButton() {
    const nextBtn = Array.from(document.querySelectorAll('a')).find(a => a.textContent.trim() === 'Next');
    if (!nextBtn) {
      console.log('❌ No Next button found');
      return false;
    }

    console.log('   ↪️ Clicking Next button...');
    nextBtn.click();
    await sleep(CONFIG.delayBetweenPages);
    return true;
  }

  function updateProgress() {
    const percent = ((state.currentPage - 1) / CONFIG.maxPages * 100).toFixed(1);
    process.stdout.write(`\r   📊 Progress: Page ${state.currentPage}/${CONFIG.maxPages} (${percent}%) | Prompts: ${state.allPrompts.length}    `);
  }

  function saveProgress() {
    const data = JSON.stringify(state.allPrompts, null, 2);
    const blob = new Blob([data], { type: 'application/json' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = `godofprompt_prompts_page_${state.currentPage}_${state.allPrompts.length}.json`;
    a.click();
    URL.revokeObjectURL(url);
  }

  // --------------------------------------------------------------------------
  // Main Extraction Logic
  // --------------------------------------------------------------------------

  console.log('📍 Navigating to page 1...');
  window.location.href = CONFIG.baseUrl;
  await sleep(4000); // Wait for initial page load

  console.log('📄 Extracting page 1...');
  let pagePrompts = extractPromptsFromCurrentPage();
  state.allPrompts.push(...pagePrompts);
  console.log(`   ✅ Page 1: Found ${pagePrompts.length} prompts\n`);

  // Main extraction loop
  for (state.currentPage = 2; state.currentPage <= CONFIG.maxPages; state.currentPage++) {
    updateProgress();

    // Try to navigate to next page
    const navigated = await clickNextButton();
    if (!navigated) {
      console.log('\n⚠️ Cannot navigate further. Stopping.');
      break;
    }

    // Extract prompts from current page
    pagePrompts = extractPromptsFromCurrentPage();

    // Check if we found new prompts
    const newOnes = pagePrompts.filter(p =>
      !state.allPrompts.some(ap => ap.url === p.url)
    );

    if (newOnes.length === 0) {
      state.emptyPageCount++;
      console.log(`   ⚠️ No new prompts on page ${state.currentPage}`);

      if (state.emptyPageCount >= CONFIG.maxEmptyPages) {
        console.log(`\n⏹️ ${state.emptyPageCount} consecutive empty pages. Stopping.`);
        break;
      }
    } else {
      state.emptyPageCount = 0;
      state.allPrompts.push(...newOnes);
      console.log(`   ✅ Page ${state.currentPage}: ${newOnes.length} new prompts (Total: ${state.allPrompts.length})`);
    }

    // Save progress every 10 pages
    if (state.currentPage % 10 === 0) {
      console.log('\n   💾 Saving checkpoint...');
      saveProgress();
    }

    // Safety stop
    if (state.allPrompts.length >= 900) {
      console.log('\n✅ Reached 900+ prompts. Near completion!');
      break;
    }
  }

  // --------------------------------------------------------------------------
  // Final Results
  // --------------------------------------------------------------------------

  console.log('\n\n' + '='.repeat(70));
  console.log('✅ EXTRACTION COMPLETE!');
  console.log('='.repeat(70));
  console.log(`📊 Total Prompts Extracted: ${state.allPrompts.length}`);
  console.log(`📄 Pages Processed: ${state.currentPage - 1}/${CONFIG.maxPages}`);
  console.log(`⏱️ Completed At: ${new Date().toLocaleString()}`);
  console.log('='.repeat(70));

  // Show sample
  console.log('\n📋 Sample Prompts (First 5):');
  state.allPrompts.slice(0, 5).forEach((p, i) => {
    console.log(`\n${i + 1}. ${p.title}`);
    console.log(`   URL: ${p.url}`);
  });

  // Final save
  console.log('\n💾 Saving final results...');
  saveProgress();

  console.log('\n✨ All done! Check your Downloads folder for the JSON file.');
  console.log('📁 Filename: godofprompt_prompts_final_' + state.allPrompts.length + '.json');

  return state.allPrompts;

})();

File: tools/godofprompt-scraper/extract_godofprompt.py
#!/usr/bin/env python3
"""
God of Prompt - Browser-based Extraction using Chrome DevTools Protocol
This script uses Selenium with CDP to monitor network requests and extract all prompts
"""

import json
import time
import csv
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

def extract_all_prompts():
    """Extract all prompts by monitoring network traffic and programmatically clicking Next"""

    # Setup Chrome with network monitoring
    chrome_options = Options()
    chrome_options.set_capability("goog:loggingPrefs", {"performance": "ALL"})
    chrome_options.add_argument("--disable-blink-features=AutomationControlled")
    chrome_options.add_argument("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")

    driver = webdriver.Chrome(options=chrome_options)

    # Storage for all prompts
    all_prompts = []
    api_endpoints = set()

    try:
        print("Loading page...")
        driver.get("https://www.godofprompt.ai/prompts?premium=false")

        # Wait for page to load
        time.sleep(5)

        # Get performance logs to find API endpoints
        logs = driver.get_log("performance")
        print(f"\nAnalyzing {len(logs)} network requests...")

        for entry in logs:
            try:
                log = json.loads(entry["message"])["message"]
                if log["method"] == "Network.requestWillBeSent":
                    url = log["params"]["request"]["url"]
                    # Look for API endpoints
                    if "api" in url.lower() or "wized" in url.lower() or "fetch" in url.lower():
                        api_endpoints.add(url)
                        print(f"Found API endpoint: {url}")
            except:
                pass

        print(f"\nFound {len(api_endpoints)} unique API endpoints")

        # Try to access Wized object in browser
        wized_script = """
        return typeof window.Wized !== 'undefined' ? {
            version: window.Wized.version,
            data: window.Wized.data,
            request: window.Wized.request,
            get: typeof window.Wized.get
        } : null;
        """

        wized_info = driver.execute_script(wized_script)
        print(f"\nWized object info: {json.dumps(wized_info, indent=2)}")

        # Get current page prompts
        prompts_script = """
        const prompts = [];
        const items = document.querySelectorAll('[wized="plp_prompt_item_all"]');
        items.forEach(item => {
            const name = item.querySelector('[wized="plp_prompt_name"]')?.textContent?.trim();
            const description = item.querySelector('[wized="plp_prompt_description"]')?.textContent?.trim();
            const id = item.querySelector('[wized="plp_prompt_id"]')?.textContent?.trim();
            if (name && name !== 'Heading') {
                prompts.push({ id, name, description });
            }
        });
        return prompts;
        """

        current_prompts = driver.execute_script(prompts_script)
        print(f"\nPage 1: Found {len(current_prompts)} prompts")
        all_prompts.extend(current_prompts)

        # Now try to navigate through pages
        max_pages = 82
        for page_num in range(2, min(6, max_pages + 1)):  # Test first 5 pages
            print(f"\nNavigating to page {page_num}...")

            # Try clicking Next button
            try:
                next_button = WebDriverWait(driver, 10).until(
                    EC.element_to_be_clickable((By.ID, "pag-next-button"))
                )
                driver.execute_script("arguments[0].click();", next_button)
                time.sleep(3)

                # Get prompts from new page
                page_prompts = driver.execute_script(prompts_script)
                print(f"Page {page_num}: Found {len(page_prompts)} prompts")
                all_prompts.extend(page_prompts)

            except Exception as e:
                print(f"Error on page {page_num}: {e}")
                break

        # Save results
        print(f"\nTotal prompts extracted: {len(all_prompts)}")

        if all_prompts:
            filename = f"godofprompt_prompts_{len(all_prompts)}_{int(time.time())}.csv"
            with open(filename, 'w', newline='', encoding='utf-8') as f:
                writer = csv.DictWriter(f, fieldnames=['id', 'name', 'description'])
                writer.writeheader()
                writer.writerows(all_prompts)
            print(f"Saved to {filename}")

            # Show sample
            print("\nSample prompts:")
            for i, prompt in enumerate(all_prompts[:3]):
                print(f"{i+1}. {prompt['name']}")
                print(f"   {prompt['description'][:100]}...")

    finally:
        driver.quit()

    return all_prompts

if __name__ == "__main__":
    print("="*60)
    print("God of Prompt - Browser-based Extraction")
    print("="*60)
    extract_all_prompts()
